File: Arrays and Dynamic Arrays/JS Code/06-maximizeExpOptimal.js
// The goal is to maximize the expression , p * arr[i]+ q*arr[j]+ r*arr[k]
// such that i < j < k and p,q,r > 0;

const input = [9, 1, 4, 2, 5, -2];

let p = 1,
	q = 2,
	r = 4;

// precompute prefixMax
const prefixMax = [...input];
prefixMax[0] = prefixMax[0] * p;
for (let i = 1; i < prefixMax.length; i++) {
	prefixMax[i] = Math.max(prefixMax[i] * p, prefixMax[i - 1]);
}

// precomput suffixMax
const suffixMax = [...input]
suffixMax[suffixMax.length - 1] = suffixMax[suffixMax.length - 1] * r;
for (let i = suffixMax.length - 2; i >= 0; i--) {
	suffixMax[i] = Math.max(suffixMax[i] * r, suffixMax[i + 1]);
}

// compute the expression
let ans = Number.MIN_SAFE_INTEGER;
for (let i = 1; i <= input.length - 2; i++) {
	let x = prefixMax[i - 1] + input[i] * q + suffixMax[i + 1];
	ans = Math.max(ans, x);
}

console.log(ans); // 37

File: Arrays and Dynamic Arrays/C++ Code/06-maximizeExpOptimal.cpp
#include<iostream>
using namespace std;
int main() {

	int n;
	cin >> n;

	vector<int> v;
	for (int i = 0; i < n; i++) {
		int x;
		cin >> x;
		v.push_back(x);
	}

	int p, q, r;
	cin >> p >> q >> r;

	vector<int> prefixMax(n);
	prefixMax[0] = v[0] * p;
	for (int i = 1; i < n; ++i)
	{
		prefixMax[i] = max(prefixMax[i - 1], v[i] * p);
	}

	vector<int> suffixMax(n);
	suffixMax[n - 1] = v[n - 1] * r;
	for (int i = n - 2; i >= 0; i--) {
		suffixMax[i] = max(suffixMax[i + 1], v[i] * r);
	}

	int ans = INT_MIN;
	for (int i = 1; i <= n - 2; ++i)
	{
		int x = prefixMax[i - 1] + v[i] * q + suffixMax[i + 1];
		ans = max(x, ans);
	}

	cout << ans << endl;
}